constantine/eth_verkle_ipa/ipa_prover.nim

```diff
@@ -48,93 +49,125 @@ func genIPAConfig*(res: var IPASettings, ipaTranscript: var IpaTranscript[Crypto
-  res.SRS.generate_random_points(ipaTranscript, uint64(VerkleDomain))
-  res.Q_val.fromAffine(Banderwagon.getGenerator())
-  res.precompWeights.newPrecomputedWeights()
-  res.numRounds.computeNumRounds(uint64(VerkleDomain))
+  var random_points: array[VerkleDomain, EC_P]
+  random_points.generate_random_points(uint64(VerkleDomain))
+  var q_val {.noInit.}: EC_P
+  q_val.fromAffine(Banderwagon.getGenerator())
+  var precomp {.noInit.}: PrecomputedWeights
+  precomp.newPrecomputedWeights()
+  var nr: uint32
+  nr.computeNumRounds(uint64(VerkleDomain))
+
+  res.numRounds = nr
+  res.Q_val = q_val
+  res.precompWeights = precomp
+  res.SRS = random_points
+
   return true
 
-func createIPAProof*[IPAProof] (res: var IPAProof, transcript: var CryptoHash, ic: IPASettings, commitment: EC_P, a: var openArray[Fr[Banderwagon]], evalPoint: Fr[Banderwagon]) : bool {.inline.} =
+func populateCoefficientVector* (res: var openArray[Fr[Banderwagon]], ic: IPASettings, point: Fr[Banderwagon])=
+  var maxEvalPointInDomain {.noInit.}: Fr[Banderwagon]
+  maxEvalPointInDomain.fromInt(VerkleDomain - 1)
+
+  for i in 0 ..< res.len:
+    res[i].setZero()
+
+  if (maxEvalPointInDomain.toBig() < point.toBig()).bool == false:
+    let p = cast[int](point.toBig())
+    res[p].setOne()
+
+  else:
+    res.computeBarycentricCoefficients(ic.precompWeights, point)
+
+
+func createIPAProof*[IPAProof] (res: var IPAProof, transcript: var CryptoHash, ic: IPASettings, commitment: EC_P, a: var openArray[Fr[Banderwagon]], evalPoint: Fr[Banderwagon]) : bool =
   ## createIPAProof creates an IPA proof for a committed polynomial in evaluation form.
   ## `a` vectors are the evaluation points in the domain, and `evalPoint` represents the evaluation point.
   transcript.domain_separator(asBytes"ipa")
-  var b {.noInit.}: array[VerkleDomain, Fr[Banderwagon]]
+  var b = newSeq[Fr[Banderwagon]](VerkleDomain)
+  b.populateCoefficientVector(ic, evalPoint)
   
-  b.computeBarycentricCoefficients(ic.precompWeights, evalPoint)
   var innerProd {.noInit.}: Fr[Banderwagon]
-
   innerProd.computeInnerProducts(a, b)
 
   transcript.pointAppend(asBytes"C", commitment)
   transcript.scalarAppend(asBytes"input point", evalPoint.toBig())
   transcript.scalarAppend(asBytes"output point", innerProd.toBig())
 
-  var w : matchingOrderBigInt(Banderwagon)
+  var w {.noInit.}: matchingOrderBigInt(Banderwagon)
   w.generateChallengeScalar(transcript, asBytes"w")
 
-  var q {.noInit.} : EC_P
+  var w_fr {.noInit.}: Fr[Banderwagon]
+  w_fr.fromBig(w)
+
+  var a_copy = newSeq[Fr[Banderwagon]](VerkleDomain)
+  for i in 0 ..< 256:
+    a_copy[i] = a[i]
+ 
+  var q {.noInit.}: EC_P
   q = ic.Q_val
   q.scalarMul(w)
 
-  var current_basis {.noInit.}: array[VerkleDomain, EC_P]
-  current_basis = ic.SRS
+  # var current_basis {.noInit.}: array[VerkleDomain, EC_P]
+  # current_basis = ic.SRS
+
+  var cb_c = newSeq[EC_P](VerkleDomain)
```

Comment:
I avoid using sequences and Nim-allocated code in general.
`allocHeapArray` should be preferred.

The previous view approach completely avoids allocations and is better.

Reply:
So here, apart from chunking `a` to `a_L` and `a_R`, and so on, I need to cut the length of `a` itself by half, during every iteration, and the IPA goes on for 8 rounds. I think when I was using `View`, probably the original array wasn't shrinking to half. Also `splitMiddle` was earlier used in `StridedView` in `kzg/research` I think but it didn't have a `toOpenArray( )` method at that point, so I went with a rather basic approach. After every iteration, if `allocHeapArray` would let me shrink the array after every iteration, then I'll definitely implement that. Could you please point out such an example?

okay I can do `View` with `chunk`, but that again can work if there's a reallocating mechanism, else it's sometimes gives me correctness issues.